src/qtum/qtumx86.cpp

```diff
@@ -371,6 +372,22 @@ uint32_t QtumHypervisor::ReadExternalStorage(uint32_t syscall, x86Lib::x86CPU& v
     return status;
 }
 
+uint32_t QtumHypervisor::SHA256(uint32_t syscall, x86Lib::x86CPU& vm) {
+    //eax = success
+    //ebx = original value
+    //ecx = sizeof original value
+    //edx = hash of original value
+    size_t len = vm.Reg32(ECX);
+    unsigned char* k = new unsigned char[len];
+    unsigned char hash[CSHA256::OUTPUT_SIZE] = {};
+    vm.ReadMemory(vm.Reg32(EBX), len, k);
+    CSHA256().Reset().Write(k, len-1).Finalize(hash); // create hash
+    vm.WriteMemory(vm.Reg32(EDX), 256, hash);
+    vm.addGasUsed(len);
```

Comment:
addGasUsed should be at the beginning of the function for security reasons. addGasUsed will cause an exception if the user has not supplied enough gas. This prevents an attack where hashing 1Gb of data could be invoked that could DoS a node

Reply:
Ah. Yea good point. Will fix.

Reply:
If this is the case, should we update the other codes to make sure that gasUsed is added at the beginning of the function?